src/cockpit/channels/filesystem.py

```diff
@@ -530,8 +530,8 @@ def do_identity_changed(self, fd: 'Handle | None', err: 'int | None') -> None:
 
     def do_close(self) -> None:
         # non-watch channels close immediately — if we get this, we're watching
-        assert self.path_watch is not None
-        self.path_watch.close()
+        if self.path_watch is not None:
```

Comment:
This conditional sort of invalidates the point of the comment above.

There's a thorny philosophical issue behind all of this: what's supposed to happen when writing to the transport fails with an `OSError`.  That's supposed to result in the transport being closed with that exception, of course, but should it also bubble up into the code?

If so, I feel like we should do more to guard individual callers against having to deal with this.  I don't want to have very channel having to think about what might happen if their `.send()` calls throw...

Reply:
Which is to say: let me do a bit of research on this and see what the standard transports do.

Reply:
```python
import asyncio
import sys
from typing import override


class Prot(asyncio.Protocol):
    @override
    def connection_made(self, transport: asyncio.BaseTransport):
        assert isinstance(transport, asyncio.WriteTransport)
        transport.write(b'hihi')
        print('I survived')

    @override
    def connection_lost(self, exc: Exception | None) -> None:
        print('lost', exc)

async def run() -> None:
    loop = asyncio.get_running_loop()
    await loop.connect_write_pipe(Prot, sys.stdin)
    await asyncio.sleep(1000)


asyncio.run(run())
```

yields

```
$ python3 x.py < /dev/null
I survived
lost [Errno 9] Bad file descriptor
```

which is to say: the transport protects the caller of `.write()` from the exception.

Reply:
Our transports should guard callers in the same way:

```python
        try:
            n_bytes = os.write(self._out_fd, data)
        except BlockingIOError:
            n_bytes = 0
        except OSError as exc:
            self.abort(exc)
            return
```

There's a critical difference, though: we call `.abort()` directly, which immediately calls `.connection_lost()` on the Protocol.

```python
    def abort(self, exc: 'Exception | None' = None) -> None:
        self._closing = True
        self._close_reader()
        self._remove_write_queue()
        self._protocol.connection_lost(exc)
        self._close()
```

The standard transports punt it to the next mainloop iteration.

```python
    def _close(self, exc=None):
        self._closing = True
        if self._buffer:
            self._loop._remove_writer(self._fileno)
        self._buffer.clear()
        self._loop._remove_reader(self._fileno)
        self._loop.call_soon(self._call_connection_lost, exc)
```

So with a standard transport, the caller's code runs first, but with ours, the abort code runs first.

The correct solution here is to change our transport to be more like the stdlib ones.  This might have also been the root cause of https://github.com/cockpit-project/cockpit/pull/20634 and other similar bugs: it's not that we close multiple times — it's that we never finished opening.

Reply:
as an aside: I remember when I was writing this transport code way at the beginning I was in a hurry and I thought "meh, I don't understand why the standard transports do it this way — it seems pointless and unnecessary.  I'll do it my way and we can change it if we find out that it's wrong."

:sweat_smile: